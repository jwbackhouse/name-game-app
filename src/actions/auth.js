import { firebase, googleAuthProvider } from '../firebase/firebase';

// NOTE this is called in app.js rather than startLogin so that it runs when app first loads, not just when user explictly logs in/out
export const login = (uid) => ({
  type:'LOGIN',
  uid
});

export const startLogin = () => {
  return () => {    // Returns a function for async actions (can be called with dispatch) - allowed via thunk
    return firebase.auth().signInWithPopup(googleAuthProvider)
    // .then((result) => {
    //   const uid = result.user.uid;
    //   dispatch(login(uid));
    // });
  };
};

export const logout = () => ({
    type:'LOGOUT',
});

export const startLogout = () => {
  return () => {
    return firebase.auth().signOut();
  };
};