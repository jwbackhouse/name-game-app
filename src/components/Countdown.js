import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import database from '../firebase/firebase';

export class Countdown extends React.Component {
  state = {
    timerOn: false,
    timerTime: 0,
    timerStart: 0,
    timerLength: this.props.game.timerLength * 1000,
    message: '',
  }
  
  componentDidMount = () => {
    // Check for start time on Firebase
    database.ref('game/startTime').on('value', snapshot => {
      const startTime = snapshot.val();
      if (startTime) {
        this.setState({ timerStart: startTime, message: '' });
        this.startTimer();
      } else {
        this.setState({ message: 'Hang on a second...' });
      }
    });
  }
  
  componentWillUnmount = () => {
    database.ref('game/startTime').off();
    this.stopTimer();
  }
  
  startTimer = () => {
    this.setState({
      timerOn: true,
    });
    
    this.timer = setInterval(() => {
      const timeLeft = this.state.timerLength - this.state.timerTime;
      if (timeLeft >= 0)  {
        this.setState({ timerTime: Date.now() - this.state.timerStart });
      } else {
        clearInterval(this.timer);
        this.setState({ timerOn: false });
        this.props.onFinished();
      }
    }, 50);
  }

  stopTimer = () => {
    clearInterval(this.timer);
    this.setState({ timerOn: false });
  };
  
  // // Unused functionality
  // resetTimer = () => {
  //   this.setState({
  //     timerTime: 0,
  //     timerStart: 0
  //   });
  // };
  
  
  render = () => {
    const { timerLength, timerTime, message } = this.state;
    const timeLeft = Math.max(0, timerLength - timerTime);
    const messageElement = message && message;
    return (
      <div>
        <p>{ messageElement }</p>
        <h3>Time left: {moment(timeLeft).format('m:ss')}</h3>
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  game: state.game,
});

export default connect(mapStateToProps)(Countdown);
