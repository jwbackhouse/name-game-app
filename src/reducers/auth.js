// *** AUTH REDUCER

const initialState = {
  uid: '',
  username: '',
  error: ''
};

export default (state = initialState, action) => {
  switch(action.type) {
    case 'LOGIN_SUCCESS':
      return {
        uid: action.uid,
        username: action.username,
        error: ''
      };
    case 'LOGIN_ERROR':
      return {
        uid: '',
        error: action.error
      };
    case 'LOGOUT_SUCCESS':
      return initialState;
    case 'LOGOUT_FAILURE':
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
