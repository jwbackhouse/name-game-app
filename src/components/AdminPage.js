import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Link, useHistory } from 'react-router-dom';
import database from '../firebase/firebase';
import withLiveData from '../helpers/withLiveData';
import { resetGame, initialiseGame } from '../actions/game';
import { removeAllNames } from '../actions/names';


const AdminPage = ({ initialiseGame, removeAllNames, resetGame, game, history }) => {
  const [submitMsg, setSubmitMsg] = useState('');
  
  // Number of passes allowed per player
  const [numPasses, setNumPasses] = useState(game.numPasses || 2);
  const onPassesChange = (e) => setNumPasses(+e.target.value);
  useEffect(() => {
    if (game.numPasses !== undefined) setNumPasses(game.numPasses)  // needed otherwise input becomes uncontrolled
  }, [game.numPasses]);
  
  // Length of timer for guesses
  const [timerLength, setTimerLength] = useState(game.timerLength || 60000);
  const onTimerChange = (e) => {
    const milliseconds = e.target.value * 1000;
    setTimerLength(milliseconds);
  };
  
  // Very basic password protection for page access
  const [access, setAccess] = useState(undefined);
  if (access === undefined ) {
    const password = prompt('Please enter your password to view this page');
    const pass1 = 'namegame';
    password === pass1 ? setAccess(true): setAccess(false);
  }
  
  const onReset = () => {
    const confirm = window.confirm('Are you sure you want to reset the game? This can\'t be undone...');
    if (confirm) {
      database.ref().remove()
        .then(() => {
          initialiseGame();
          setSubmitMsg('Data successfully reset.');
        });
      removeAllNames();
      resetGame();
    }
  }
  
  const onSubmit = (e) => {
    e.preventDefault();
    
    database.ref('game').update({
      numPasses,
      timerLength,
    });
    setSubmitMsg('All saved for you.');
  };
  
  const pageContent = (
    <div>
      <h1>Game settings</h1>
      <form onSubmit={ onSubmit }>
        <label>Number of passes allowed
          <input
            name='numPasses'
            id='numPasses'
            value={ numPasses }
            onChange={ onPassesChange }
            type='number'
          />
        </label>
        <label>Time allowed for guessing (in seconds)
          <input
            name='timerLength'
            id='timerLength'
            value={ timerLength / 1000 }
            onChange={ onTimerChange }
            type='number'
          />
        </label>
        <button type='submit'>
          Save
        </button>
      </form>
      { submitMsg && <p>{ submitMsg }</p> }
      <br />
      <button onClick={ onReset }>
        Reset the game
      </button>
      <br />
      <br />
      <Link
        to='/'
        className='button'
      >
        Go home
      </Link>
    </div>
  );
  
  return access ? pageContent : null;
};

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  removeAllNames: () => dispatch(removeAllNames()),
  resetGame: () => dispatch(resetGame()),
  initialiseGame: () => dispatch(initialiseGame())
});


const connectedWithLiveData = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLiveData,
);

export default connectedWithLiveData(AdminPage);
