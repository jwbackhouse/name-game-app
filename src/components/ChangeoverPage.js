import React from 'react';
import { connect } from 'react-redux';
import StartMessage from './StartMessage';
import { fetchData, endFetchData, startTurn } from '../actions/game';
import { getNames } from '../actions/names';
import selectPlayer from '../selectors/selectPlayer';



export class ChangeoverPage extends React.Component {
  state = {
    error: '',
    nextPlayer: undefined
  }
  
  componentDidMount = () => {
    this.props.fetchData();
    this.props.getNames();
  }
  
  componentDidUpdate = (prevProps) => {
    if (!prevProps.game.startTurn) {
      this.props.game.startTurn && this.props.history.push('/guess');
    }
    if (this.props.game.endGame) {
      this.props.history.push('/end')
    }
  }
  
  componentWillUnmount = () => {
    this.props.endFetchData();
  }

  onClick = () => {
    // Update Firebase with startGame & startTime
    this.props.startTurn();
    this.props.history.push('/play')
  }
  
  render = () => {
    // Check if this user is playing next
    const thisUserPlaying = this.props.user.uid === this.props.game.playingNow.uid;

    return (
      <div className='content-container'>
        <h1>The scores on the doors</h1>
        <p>Team A: {this.props.game.teamAScore}</p>
        <p>Team B: {this.props.game.teamBScore}</p>
        <StartMessage
          thisUserPlaying={ thisUserPlaying }
          playingNow={ this.props.game.playingNow }
          errorMsg={ '' }
          onClick={ this.onClick }
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  players: state.players,
  game: state.game,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  startTurn: () => dispatch(startTurn()),
  fetchData: () => dispatch(fetchData()),
  getNames: () => dispatch(getNames()),
  endFetchData: () => dispatch(endFetchData())
});


export default connect(mapStateToProps, mapDispatchToProps)(ChangeoverPage);

