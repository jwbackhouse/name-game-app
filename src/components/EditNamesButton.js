import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { Link } from 'react-router-dom';
import { togglePlayerReady } from '../actions/players';

export const EditNamesButtonBase = props => {
  const onTogglePlayerReady = () => {
    props.togglePlayerReady(props.auth.uid, props.auth.firebaseUID);
    props.history.push('/setup');
  };
  
  return (
    <button className='button button--background' onClick={ onTogglePlayerReady }>
      Edit your names
    </button>
  );
};

const mapStateToProps = state => ({
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  togglePlayerReady: (uid, playersUid) => dispatch(togglePlayerReady(uid, playersUid)),
})

// withRouter required to access history
const EditNamesButton = withRouter(EditNamesButtonBase);
export default connect(mapStateToProps, mapDispatchToProps)(EditNamesButton);
