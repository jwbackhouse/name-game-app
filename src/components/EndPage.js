import React from 'react';
import { connect } from 'react-redux';
import database from '../firebase/firebase';
import { fetchData } from '../actions/game';
import { removeAllNames } from '../actions/names';


export class EndPage extends React.Component {
  componentDidMount = () => {
    this.props.fetchData();
  }
    
  onClick = () => {
    // Clear database
    database.ref().remove();
    // Necessary as db update on Register page doesn't over-write names
    this.props.removeAllNames();
    this.props.history.push('/');
  }
  
  render = () => {
    let message;
    if (this.props.game.teamAScore > this.props.game.teamBScore) {
      message = 'Team A has it!';
    } else if (this.props.game.teamBScore > this.props.game.teamAScore) {
      message = 'Team B has it!';
    } else {
      message = "It's a draw!";
    }
    
    return (
      <div className='content-container'>
        <h1>{ message }</h1>
        <p>Team A scored { this.props.game.teamAScore }</p>
        <p>Team B scored { this.props.game.teamBScore }</p>
        <button onClick={ this.onClick }>Play again</button>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  fetchData: () => dispatch(fetchData()),
  removeAllNames: () => dispatch(removeAllNames())
});

export default connect(mapStateToProps, mapDispatchToProps)(EndPage);

