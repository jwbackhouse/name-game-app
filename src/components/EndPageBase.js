import React from 'react';
import { connect } from 'react-redux';
import database from '../firebase/firebase';
import EndPage from './EndPage';
import { resetGame, initialiseGame, fetchScores } from '../actions/game';
import { removeAllNames } from '../actions/names';


export class EndPage extends React.Component {
  state = {
    ready: false
  }
  
  async componentDidMount() {
    console.log('componentDidMount. state.ready is', this.state.ready);
    await this.props.fetchScores();
    this.setState({ ready: true });
    console.log('fetchScores executed successfully. state.ready is', this.state.ready);
  }
    
  onClick = () => {
    // Clear database & initialise new game
    database.ref().remove()
      .then(() => this.props.initialiseGame());
    // Necessary as db update on Register page doesn't over-write names / all game data
    this.props.removeAllNames();
    this.props.resetGame();
    this.props.history.push('/');
  }
  
  render = () => {
    let message;
    if (this.props.game.teamAScore > this.props.game.teamBScore) {
      message = 'Team A has it!';
    } else if (this.props.game.teamBScore > this.props.game.teamAScore) {
      message = 'Team B has it!';
    } else {
      message = "It's a draw!";
    }
    
    const renderComponent = this.state.ready && (
      <EndPage
          message={ message }
          teamAScore={ this.props.game.teamAScore }
          teamBScore={ this.props.game.teamBScore }
          onClick={ this.onClick }
      />
    )
    
    return (
      <span>{ renderComponent }</span>
    )
  }
}

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  fetchData: () => dispatch(fetchData()),
  fetchScores: () => dispatch(fetchScores()),
  removeAllNames: () => dispatch(removeAllNames()),
  resetGame: () => dispatch(resetGame()),
  initialiseGame: () => dispatch(initialiseGame())
});

export default connect(mapStateToProps, mapDispatchToProps)(EndPage);

