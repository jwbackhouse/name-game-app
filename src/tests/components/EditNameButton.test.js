import React from 'react';
import { shallow } from 'enzyme';
import { EditNamesButtonBase } from '../../components/EditNamesButton';
import auth from '../fixtures/testAuth';

describe('<EditNamesButton />', () => {
  let wrapper;
  const togglePlayerReady = jest.fn();
  const history = { push: jest.fn() };
  beforeEach(() => {
    wrapper = shallow(
      <EditNamesButtonBase
        togglePlayerReady={ togglePlayerReady }
        history={ history }
        auth={ auth }
      />)
  })
  
  test('Should render EditNamesButton', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('Should call togglePlayerReady on click and redirect to /setup', () => {
    wrapper.find('button').simulate('click');
    expect(togglePlayerReady).toBeCalled();
    expect(history.push).toBeCalledWith('/setup');
  });
});
