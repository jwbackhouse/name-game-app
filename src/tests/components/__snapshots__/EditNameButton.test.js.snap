// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditNamesButton /> Should render EditNamesButton 1`] = `
<button
  onClick={[Function]}
>
  Edit your names
</button>
`;
